Vekterli/make clustercontroller state generation pure functional rebased part deux

vekterli opened the pull request:
@hakonhall This is a big one. For the first pass, I'd recommend focusing on `ClusterStateGenerator` and `ClusterStateGeneratorTest`.

Lots of TODOs and FIXMEs still sprinkled around; I hope to address as many of these as possible this week.

No need to look at individual commits; tons of refactoring happening between them.

vekterli committed:
Add support for non-temporal (reported/wanted) states

vekterli committed:
Add transient maintenance mode on down-edge for storage nodes

vekterli committed:
Handle startup timestamp propagation and node up-count limits

vekterli committed:
Wire in group auto take-down feature in new state generator

vekterli committed:
Compute distribution bits based on config and current node min bits

vekterli committed:
Begin adding separate component for emitting node/cluster change events

vekterli committed:
Emit events for groups up/down edges per node

vekterli committed:
Emit events for cluster down edges with explicit reasons

vekterli committed:
Test and handle emitting events with correct type and time

vekterli committed:
Annotate generated cluster states with reasons

vekterli committed:
Handle unstable init cases

Currently code for auto-crosschecking with legacy state generator
has been added. MasterElectionTest fails here for yet unknown reasons.

vekterli committed:
Add conversion from controller options to generator params

vekterli committed:
Begin integration of new code with the legacy controller

vekterli committed:
Must force state version publish after new version fetched from ZK

vekterli committed:
Begin removing and refactoring old code

vekterli committed:
More refactoring. Tests currently broken.

vekterli committed:
Wire in orchestration tracking components and new version history

vekterli committed:
Tear out more (now dead) legacy code

vekterli committed:
Refactoring

vekterli committed:
Factor out reconfig to own method

vekterli committed:
Fix consolidated cluster state regression

vekterli committed:
Remove duplicate and dead test code

vekterli committed:
Timer events must use consolidated cluster state, not published

vekterli committed:
Remove now redundant branch in StateChangeHandler

vekterli committed:
More dead code removal and un-breakage of test

vekterli committed:
Implement separate similarity check which ignores init progress

hakonhall reviewed (commented):
I have mainly looked at ClusterStateGenerator

vekterli committed:
Use Optional instead of passing/storing null

vekterli committed:
Move logic for implicitly marking nodes listing buckets as Down

vekterli committed:
s/upper bound/lower bound/ for future distribution bit persistence work

vekterli committed:
Make time unit explicit from naming

vekterli committed:
Use baseline state instead of reported state to decide Wanted state overrides

vekterli committed:
Test returned cluster state reason for a fully downed cluster

Currently reports as too few storage nodes available, rather than a
specific "all nodes are down" reason. This can be changed at any point.

vekterli committed:
More simplification of equals/hashCode

hakonhall reviewed (commented):
From going through the rest of the PR...

vekterli committed:
Move emptyState utility factory method into base ClusterState class

vekterli committed:
Simplify proposeNewNodeState

vekterli committed:
Refactoring, less "always on" log string construction

vekterli committed:
Move remaining ClusterStateUtil functionality into ClusterState

vekterli committed:
Disable max init progress feature if configured value is zero

vekterli committed:
Test that maintenance nodes are not affected by implicit group down edges

vekterli committed:
Factor out reported state logic into smaller functions

vekterli committed:
Update test comments re: disk states for nodes in down state

vekterli committed:
Remove leftover debug logging

vekterli committed:
Test that simply setting a non-zero crash count does not mark a node as unstable

vekterli committed:
Clarify when toString() representation is used for state history entries

vekterli committed:
Only process a single remote task per tick invocation

Avoids situation where current cluster state might be mutated by an operation,
but where this change would not be visible to subsequent tasks due to these
sharing the same processing context.
